Replace _WKNodeInfo with _WKJSHandle

achristensen07 opened the pull request:
#### 6ea6f93a63148e359620e0dd74e4b7720ddc6b0a
<pre>
Replace _WKNodeInfo with _WKJSHandle
<a href="https://bugs.webkit.org/show_bug.cgi?id=297254">https://bugs.webkit.org/show_bug.cgi?id=297254</a>

Reviewed by Ryosuke Niwa.

_WKNodeInfo was a good prototype.  It gave us a way to reference a node in the UI process
and pass it in as an argument to WKWebView.callAsyncJavaScript.

However, in using it we have found some limitations.  It can only get a WKFrameInfo if the
node is an HTMLFrameOwnerElement, so there is no way to get the WKFrameInfo of window.parent.
Also, we can only pass a node as an argument to a JS function.  There are lots of objects
we want to pass in that aren&apos;t nodes, like other JS functions.

To meet these needs and make a slightly more powerful primitive, I replace it with _WKJSHandle.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/page/WebKitNamespace.cpp:
(WebCore::WebKitNamespace::createWeakRef):
(WebCore::WebKitNamespace::createNodeInfo): Deleted.
* Source/WebCore/page/WebKitNamespace.h:
* Source/WebCore/page/WebKitNamespace.idl:
* Source/WebCore/page/WebKitWeakRef.cpp: Added.
(WebCore::weakRefMap):
(WebCore::documentMap):
(WebCore::WebKitWeakRef::weakRefDestroyed):
(WebCore::windowFrameIdentifier):
(WebCore::WebKitWeakRef::WebKitWeakRef):
* Source/WebCore/page/WebKitWeakRef.h: Renamed from Source/WebCore/page/WebKitNodeInfo.h.
(WebCore::WebKitWeakRef::create):
(WebCore::WebKitWeakRef::identifier const):
(WebCore::WebKitWeakRef::windowFrameIdentifier const):
* Source/WebCore/page/WebKitWeakRef.idl: Renamed from Source/WebCore/page/WebKitNodeInfo.idl.
* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::toAPI):
(WebKit::roundTripThroughSerializedScriptValue):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::JavaScriptEvaluationResult::toJS):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::toID):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::isSerializable):
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in:
* Source/WebKit/Shared/NodeHitTestResult.h:
* Source/WebKit/Shared/NodeHitTestResult.serialization.in:
* Source/WebKit/Shared/ProcessQualified.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WeakRefInfo.h: Renamed from Source/WebKit/Shared/NodeInfo.h.
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIWeakRef.cpp: Copied from Source/WebKit/UIProcess/API/APINodeInfo.cpp.
(API::WeakRef::WeakRef):
(API::WeakRef::~WeakRef):
* Source/WebKit/UIProcess/API/APIWeakRef.h: Renamed from Source/WebKit/UIProcess/API/APINodeInfo.h.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _hitTestAtPoint:inFrameCoordinateSpace:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKNodeInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKNodeInfo.mm:
(-[_WKNodeInfo contentFrameInfo:]):
(-[_WKNodeInfo dealloc]): Deleted.
(-[_WKNodeInfo _apiObject]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKWeakRef.h: Renamed from Source/WebKit/UIProcess/API/APINodeInfo.cpp.
* Source/WebKit/UIProcess/API/Cocoa/_WKWeakRef.mm: Renamed from Source/WebCore/page/WebKitNodeInfo.cpp.
(-[_WKWeakRef dealloc]):
(-[_WKWeakRef windowFrameInfo:]):
(-[_WKWeakRef _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWeakRefInternal.h: Renamed from Source/WebKit/UIProcess/API/Cocoa/_WKNodeInfoInternal.h.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::hitTestAtPoint):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::weakRefDestroyed):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SerializedNode.mm: Renamed from Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm.
(TestWebKitAPI::TEST(SerializedNode, Basic)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, HitTesting)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WeakRef.mm: Added.
(TestWebKitAPI::getWindowFrameInfo):
(TestWebKitAPI::TEST(WeakRef, Basic)):

Canonical link: <a href="https://commits.webkit.org/298585@main">https://commits.webkit.org/298585@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5c3fa8b5afb6e35f4acca4377df1eff22483657a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115975 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35629 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26165 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/122024 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66511 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/4af38ae1-d4e4-4e1e-8e9a-57a74b57e99b) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117864 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36318 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44216 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88098 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42685 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/d181d3ae-2e8a-4e11-82c6-a8e2cd704048) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118923 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28974 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104066 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68508 "Passed tests") | | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/115349 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28098 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22175 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65694 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98363 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22309 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125177 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42862 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32174 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96855 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43228 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100254 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96639 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41899 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19772 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38791 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/18532 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42751 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48348 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42218 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45552 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43925 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

rniwa commented on Source/WebCore/CMakeLists.txt:
```diff
@@ -1559,9 +1559,9 @@ set(WebCore_NON_SVG_IDL_FILES
-    page/WebKitNodeInfo.idl
     page/WebKitPoint.idl
     page/WebKitSerializedNode.idl
+    page/WebKitWeakRef.idl
```
`WebKitWeakRef` sounds awfully confusing with `WeakRef`.
I think we wanna say that this is a reference to a remote object in WebContent process.
Maybe something like `WKWeakContentRef` or `WKRemoteObjectRef`?

achristensen07 replied:
WKObjectProxy?

rniwa replied:
Hm... we don't use `Proxy` in a public API, do we? Maybe `WKContentObject`?

achristensen07 replied:
Brady and I discussed and like _WKJSHandle. It's not final and can be renamed if promoted to API.

achristensen07 committed:
Replace _WKNodeInfo with _WKJSHandle
https://bugs.webkit.org/show_bug.cgi?id=297254

Reviewed by Ryosuke Niwa.

_WKNodeInfo was a good prototype.  It gave us a way to reference a node in the UI process
and pass it in as an argument to WKWebView.callAsyncJavaScript.

However, in using it we have found some limitations.  It can only get a WKFrameInfo if the
node is an HTMLFrameOwnerElement, so there is no way to get the WKFrameInfo of window.parent.
Also, we can only pass a node as an argument to a JS function.  There are lots of objects
we want to pass in that aren't nodes, like other JS functions.

To meet these needs and make a slightly more powerful primitive, I replace it with _WKJSHandle.

* Source/WebCore/CMakeLists.txt:
* Source/WebCore/DerivedSources-input.xcfilelist:
* Source/WebCore/DerivedSources-output.xcfilelist:
* Source/WebCore/DerivedSources.make:
* Source/WebCore/Headers.cmake:
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/bindings/js/WebCoreBuiltinNames.h:
* Source/WebCore/page/WebKitNamespace.cpp:
(WebCore::WebKitNamespace::createWeakRef):
(WebCore::WebKitNamespace::createNodeInfo): Deleted.
* Source/WebCore/page/WebKitNamespace.h:
* Source/WebCore/page/WebKitNamespace.idl:
* Source/WebCore/page/WebKitWeakRef.cpp: Added.
(WebCore::weakRefMap):
(WebCore::documentMap):
(WebCore::WebKitWeakRef::weakRefDestroyed):
(WebCore::windowFrameIdentifier):
(WebCore::WebKitWeakRef::WebKitWeakRef):
* Source/WebCore/page/WebKitWeakRef.h: Renamed from Source/WebCore/page/WebKitNodeInfo.h.
(WebCore::WebKitWeakRef::create):
(WebCore::WebKitWeakRef::identifier const):
(WebCore::WebKitWeakRef::windowFrameIdentifier const):
* Source/WebCore/page/WebKitWeakRef.idl: Renamed from Source/WebCore/page/WebKitNodeInfo.idl.
* Source/WebKit/Modules/OSX_Private.modulemap:
* Source/WebKit/Modules/iOS_Private.modulemap:
* Source/WebKit/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebKit/Scripts/webkit/messages.py:
(serialized_identifiers):
(types_that_cannot_be_forward_declared):
(headers_for_type):
* Source/WebKit/Scripts/webkit/tests/MessageArgumentDescriptions.cpp:
(IPC::serializedIdentifiers):
* Source/WebKit/Shared/API/APIObject.h:
* Source/WebKit/Shared/Cocoa/APIObject.mm:
(API::Object::newObject):
* Source/WebKit/Shared/JavaScriptEvaluationResult.cpp:
(WebKit::JavaScriptEvaluationResult::toAPI):
(WebKit::roundTripThroughSerializedScriptValue):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::JavaScriptEvaluationResult::toJS):
* Source/WebKit/Shared/JavaScriptEvaluationResult.h:
* Source/WebKit/Shared/JavaScriptEvaluationResult.mm:
(WebKit::JavaScriptEvaluationResult::toID):
(WebKit::JavaScriptEvaluationResult::toValue):
(WebKit::isSerializable):
* Source/WebKit/Shared/JavaScriptEvaluationResult.serialization.in:
* Source/WebKit/Shared/NodeHitTestResult.h:
* Source/WebKit/Shared/NodeHitTestResult.serialization.in:
* Source/WebKit/Shared/ProcessQualified.serialization.in:
* Source/WebKit/Shared/WTFArgumentCoders.serialization.in:
* Source/WebKit/Shared/WeakRefInfo.h: Renamed from Source/WebKit/Shared/NodeInfo.h.
* Source/WebKit/Sources.txt:
* Source/WebKit/SourcesCocoa.txt:
* Source/WebKit/UIProcess/API/APIWeakRef.cpp: Copied from Source/WebKit/UIProcess/API/APINodeInfo.cpp.
(API::WeakRef::WeakRef):
(API::WeakRef::~WeakRef):
* Source/WebKit/UIProcess/API/APIWeakRef.h: Renamed from Source/WebKit/UIProcess/API/APINodeInfo.h.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _hitTestAtPoint:inFrameCoordinateSpace:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebViewPrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKNodeInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKNodeInfo.mm:
(-[_WKNodeInfo contentFrameInfo:]):
(-[_WKNodeInfo dealloc]): Deleted.
(-[_WKNodeInfo _apiObject]): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKWeakRef.h: Renamed from Source/WebKit/UIProcess/API/APINodeInfo.cpp.
* Source/WebKit/UIProcess/API/Cocoa/_WKWeakRef.mm: Renamed from Source/WebCore/page/WebKitNodeInfo.cpp.
(-[_WKWeakRef dealloc]):
(-[_WKWeakRef windowFrameInfo:]):
(-[_WKWeakRef _apiObject]):
* Source/WebKit/UIProcess/API/Cocoa/_WKWeakRefInternal.h: Renamed from Source/WebKit/UIProcess/API/Cocoa/_WKNodeInfoInternal.h.
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebProcess/WebPage/WebPage.cpp:
(WebKit::WebPage::hitTestAtPoint):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebProcess.cpp:
(WebKit::WebProcess::weakRefDestroyed):
* Source/WebKit/WebProcess/WebProcess.h:
* Source/WebKit/WebProcess/WebProcess.messages.in:
* Tools/TestWebKitAPI/SourcesCocoa.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SerializedNode.mm: Renamed from Tools/TestWebKitAPI/Tests/WebKitCocoa/NodeInfo.mm.
(TestWebKitAPI::TEST(SerializedNode, Basic)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/SiteIsolation.mm:
(TestWebKitAPI::(SiteIsolation, HitTesting)):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WeakRef.mm: Added.
(TestWebKitAPI::getWindowFrameInfo):
(TestWebKitAPI::TEST(WeakRef, Basic)):

Canonical link: https://commits.webkit.org/298585@main
